design/nfs/nfs-controller-runtime.md

```diff
@@ -0,0 +1,161 @@
+
+## Background
+
+This proposal is to implement controller-runtime in Rook NFS Operator to improve reliability of the operator itself. Currently, Rook nfs-operator only simply watches an event of CustomResource from an informer using simple WatchCR method which has limited functionality such as event can not be re-queued if failed. To implement controller-runtime is expected to overcome the shortcomings of current implementation.
+
+## Why controller-runtime?
+
+Controller-runtime is widely used for writing Kubernetes operators. It is also leveraged by Kubebuilder and Operator SDK. Controller-runtime consists of several packages that have their respective responsibilities in building operators. The main function of controller-runtime is
+
+- **Manager:** Runnable for the operator with leader election option. It is also provides shared dependencies such as clients, caches, schemes, etc.
+- **Controller:** Provides types and functions for building Controllers which ensure for any given object, the actual state matches the desired state which called `Reconciling` process.
+- **Admission Webhook:** Provides methods to build an admission webhook (both Mutating Admission Webhook and Validating Admission Webhook) and bootstrap a webhook server.
+- **Envtest:** Provides libraries for integration testing by starting a local control plane (etcd and kube-apiserver).
+- **Matrics:** Provides metrics utility for controller.
+
+## Implementation
+
+The implementation of this proposal is to rewrite NFS Operator controller to use controller-runtime and introduce the validation admission webhook using controller-runtime for NFS Operator.
+
+### Controller & Reconciliation
+
+Operators are Kubernetes extensions that use custom resources to manage applications and their components using the Kubernetes APIs and kubectl tooling. Operators follow the Kubernetes controller principles. The process in which the actual state of the object (both cluster object and external object) will be matching the desired state which called *Reconciliation* process in the controller-runtime. 
+
+The current implementation is the operator watch an event (create, update and delete) of CustomResource and will be handled by registered function in `ResourceEventHandlerFuncs` which every event has its own handler but only the create handler that implemented.
+
+Controller-runtime introduces an interface called `Reconciler` that will ensure the state of the system matches what is specified by the user in the object at the time the `Reconciler` is called. `Reconciler` responds to generic events so it will contain all of the business logic of a Controller (create, update, and delete). The controller-runtime also have utility functions for creating and updating an object called `CreateOrUpdate` which will make easier to handling update of an object.
+
+Since this implementation only changes the logic of the controller, so the deployment process will be like current implementation.
```

Comment:
This is good to know that the reconciler logic doesn't require any deployment experience changes.  The same isn't true for the webhook stuff, so maybe we can clarify this sentence a bit:

> Since this controller-runtime based reconciler implementation only changes the logic of the controller, the deployment process for the basic NFS operator will be like the current implementation.  If the user chooses to also install the optional validation webhook described in the next section, the deployment process will contain new steps to follow.